src/tools.rs

```diff
@@ -100,14 +101,16 @@ pub fn extract_string(raw: &[u8], off: usize) -> Option<&str> {
     if off >= raw.len() {
         return None;
     }
-    while raw[end] != 0 {
+    while end < raw.len() && raw[end] != 0 {
         end += 1;
     }
-    let blk = raw[off..end].as_ptr() as *mut u8;
-    let r = unsafe { String::from_raw_parts(blk, end - off, end - off) };
-    let ret = Some(unsafe { &*(r.as_str() as *const str) }); // eliminate lifetime
-    r.into_bytes().leak();
-    ret
+    if end >= raw.len() {
+        return None;
+    }
+    end += 1;
+    let blk = raw[off..end].as_ptr() as *const i8;
+    let r = unsafe { CStr::from_ptr(blk) };
+    r.to_str().ok()
```

Comment:
I think this can be shortened to:
```rust
pub fn extract_string(raw: &[u8], off: usize) -> Option<&str> {
    if off >= raw.len() {
        return None;
    }
    CStr::from_bytes_with_nul(&raw[off..]).ok()?.to_str().ok()
}
```
Right?

Reply:
Actually the NUL byte detection is still necessary, because `from_bytes_with_nul` assumes no interior NUL bytes. But we don't have to use `unsafe` at least and it's shorter still.

Reply:
Got it. `from_bytes_with_nul()` is better.